src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -499,352 +527,305 @@ public Task<Compilation> GetCompilationAsync(SolutionCompilationState compilatio
+                        Contract.ThrowIfTrue(initialState.IntermediateProjects.IsEmpty);
 
-                    // This is guaranteed by an in progress state.  Which means we know we'll get into the while loop below.
-                    Contract.ThrowIfTrue(intermediateProjects.Count == 0);
-                    AllSyntaxTreesParsedState? resultState = null;
+                        WithCompilationTrackerState currentState = initialState;
+                        while (currentState is InProgressState inProgressState)
+                        {
+                            cancellationToken.ThrowIfCancellationRequested();
+
+                            // We have a list of transformations to get to our final compilation; take the first transformation and apply it.
+                            var (compilationWithoutGeneratedDocuments, staleCompilationWithGeneratedDocuments, generatorInfo) =
+                                await ApplyFirstTransformationAsync(inProgressState).ConfigureAwait(false);
+
+                            // We have updated state, so store this new result; this allows us to drop the intermediate state we already processed
+                            // even if we were to get cancelled at a later point.
+                            //
+                            // As long as we have intermediate projects, we'll still keep creating InProgressStates.  But
+                            // once it becomes empty we'll produce an AllSyntaxTreesParsedState and we'll break the loop.
+                            //
+                            // Preserve the current frozen bit.  Specifically, once states become frozen, we continually make
+                            // all states forked from those states frozen as well.  This ensures we don't attempt to move
+                            // generator docs back to the uncomputed state from that point onwards.  We'll just keep
+                            // whateverZ generated docs we have.
+                            currentState = CompilationTrackerState.Create(
+                                inProgressState.IsFrozen,
+                                compilationWithoutGeneratedDocuments,
+                                generatorInfo,
+                                staleCompilationWithGeneratedDocuments,
+                                inProgressState.IntermediateProjects.RemoveAt(0));
+                            this.WriteState(currentState);
+
+                            Contract.ThrowIfTrue(inProgressState.IntermediateProjects.Count > 1 && currentState is not InProgressState);
+                            Contract.ThrowIfTrue(inProgressState.IntermediateProjects.Count == 1 && currentState is not AllSyntaxTreesParsedState);
+                        }
 
-                    while (!intermediateProjects.IsEmpty)
+                        Contract.ThrowIfTrue(currentState is not AllSyntaxTreesParsedState);
```

Comment:
```suggestion
```

Your cast on the next line will enforce this anyways, and the InvalidCastException will say what the wrong type was, instead of leaving us guessing.

Reply:
fair.  note: i clean this up in a later PR where i get rid of AllSyntaxTreesParsedState